RFC Draft.md

```diff
@@ -382,9 +382,249 @@ const uploadedObjectReference = await uploadTask.result;
+
+**Current Usage (v5)**
+
+```typescript
+const todoDetails: CreateTodoInput = {
+  name: 'Todo 1',
+  description: 'Learn AWS AppSync',
+};
+
+const newTodo = await API.graphql<GraphQLQuery<CreateTodoMutation>>({
+  query: mutations.createTodo,
+  variables: { input: todoDetails },
+});
+
+const subscription = API.graphql<GraphQLSubscription<OnCreateTodoSubscription>>(
+  graphqlOperation(subscriptions.onCreateTodo)
+).subscribe({
+  next: ({ provider, value }) => console.log({ provider, value }),
+  error: (error) => console.warn(error),
+});
+```
+
+**Proposed Usage (v6)**
+
+```typescript
+type MyQueryType = {
+  result: {
+    id: string;
+    name: string;
+    description: string;
+  };
+};
+
+const result = await API.query<MyQueryType>('query getTodo...', {
+  variables: { ... }
+});
+
+console.log(
+  `Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`
+);
+
+type MyMutationType = {
+  variables: {
+    id: number;
+    name: string;
+    description: string;
+  };
+  result: {
+    id: number;
+    name: string;
+    description: string;
+  };
+};
+
+const result = await API.mutate<MyMutationType>('mutation createTodo....', {
+  variables: {
+    id: 123,
+    name: 'My Todo',
+    description: 'This is a todo',
+  },
+});
+
+console.log(
+  `Todo : ${result.data?.id}: ${result.data?.name} (${result.data?.description})`
+);
+
+API.subscribe<MyQueryType>('subscription OnCreateTodo...').on({
+  next: (result) =>
+    console.log(
+      `Todo info: ${result.data?.id}: ${result.data?.name} (${result.data?.description})`
+    ),
+});
+```
+
+## Less verbose type definitions for generated queries, mutations, and subscriptions
+In v6, we want to reduce the verbosity of the typings for the code-generated queries, mutations, and subscriptions by inferring their types from the generated code.
+
+**Current Usage (v5)**
+```ts
+import { API, graphqlOperation } from 'aws-amplify';
+import { GraphQLQuery, GraphQLSubscription } from '@aws-amplify/api';
+import { createTodo } from './graphql/mutations';
+import { onCreateTodo } from './graphql/subscriptions';
+import {
+  CreateTodoInput,
+  CreateTodoMutation,
+  OnCreateTodoSubscription,
+} from './API';
+// so many imports from disparate modules 
+
+function createMutation() {
+  const createInput: CreateTodoInput = {
+    name: 'Improve API TS support ',
+  };
+
+  // Verbose explicit type definition when the information could be available in `createTodo`'s type
+  const res = await API.graphql<GraphQLQuery<CreateTodoMutation>>(
+    graphqlOperation(createTodo, {
+      input: createInput,
+    })
+  );
+
+  // the returned data is nested 2 levels deep and could be upleveled when the GraphQL document
+  // only includes one query or mutation
+  const newTodo = res.data?.createTodo;
+
+  console.log('new todo created', newTodo);
```

Comment:
IMO we can leave out the console.logs unless we're using it to show something about the structure of the result